simpleviewinc/keg-hub
# Zr 354 classname updates

**Ticket [ZEN-354](https://jira.simpleviewtools.com/browse/ZEN-354)** 

**Notes**
* Originally this ticket was to add a `dataSet.class` attribute to the components
  * This has been changed to instead use the `className` prop
  * This in more consistent with other web platforms
  * On Native, it is ignored

## Context
* Events-Force is defining their styles in CSS
* Their clients can define custom CSS as well
* Currently, the tap, using re-theme, is only compatible with JSON style objects
* We need to be able to pull styles from stylesheets already applied to the DOM and apply them to keg-components

## Goal
* Update all keg-components to accept a className prop
  * Using the `useClassname` and `useClasslist` hooks, the `className` can be used instead of the `dataSet`
* Add a default `className` prop to all primitive components
* Merge the passed in `className` prop with the default `className` prop


## Updates
**Global**
  * Removed `@expo/vector-icons` && `react-native-vector-icons`
  * Added `react-native-svg` && `react-native-svg-web`
  * Added `svgr cli` to convert SVG's to React Native Components
  * Added scripts helpers to do SVG conversion
  * Removed all dataSet references
  * Updated alias / extentions to work like Re-Theme does
* `repos/keg-components/README.md`
  * Updated with more information
  * Added notes about updated / added code
* `repos/keg-components/configs/aliases.json`
  * Cleaned up to match code updates
* `repos/keg-components/configs/jest.config.js`
  * Removed `@expo/vector-icons`
* `repos/keg-components/setupTests.js`
  * Added mocks for `react-native-svg` to `react-native-svg-web`
  * Removed a number of the `<component>.wrapper.js ` as they are no longer needed

**Assets/Icons**
* `repos/keg-components/src/assets/icons/**`
  * Added a number of svg icons as react components
    * Mostly used in stories, but also used in some places of the code
* `repos/keg-components/src/assets/icons/svgIcon/svgIcon.js`
  * Added SVG component for consistent SVG icon rendering
* `repos/keg-components/src/assets/spinners/**`
  * Removed spinners as they are no longer being used

**Components**
* `repos/keg-components/src/components/**`
  * Updated to use `useClassList` and `useClassName` hooks as needed
  * Updated to use `SVG` component as needed
  * Added a **Touchable** component
    * Replaced components that were using `TouchableOpacity` and `TouchableNativeFeedback`
  * Added **accessibilityRole** to components where needed
    * This helps with symantec HTML, as well Dom element type to be used 
  * Removed Component wrapper in cases where no longer needed
    * `Button` / `Form` components
  * Added `createElement.js` component
    * Allows creating custom Components using the `react-native-web` framework
  * Fixed `Checkbox` component styles
    * Added styles are defined here so that they are enforced
    * Due to how the checkbox is designed, these styles should not be changed 
    * Which is why they are not defined within the theme
  * Added helpers to `noop`
    * `noPropObj` && `noPropArr`
    * Replace locations in components where they were just being set to an empty object or array
  * Fixed bug with TouchableIcon causing the wrapper styles to not be used
  * Updated `style` prop in a number of locations to use the `array` syntax
  * `repos/keg-components/src/components/box/textBox.js`
    * Fixed issue with icon styles to being applied properly
* `repos/keg-components/src/components/card/card.js`
  * Lots of updates to clean this up
  * Added more stories
* `repos/keg-components/src/components/drawer/drawer.js`
  * Added this but have not full set it up yet
  * I'm using it in one of my own taps, so figured I copy it over to here
  * I'll updating it more in the next pull request
  * It's not being exported, so does not effect anything
* `repos/keg-components/src/components/form/select/picker.js`
  * Same as the `drawer` component
  * Needs a lot more work, but this is a start
  * It's not being exported, so does not effect anything
* `repos/keg-components/src/components/image/image.js`
  * Lots of clean up to fix how Images and loading works
    * Now uses the Loading component properly
    * Added some custom code here to ensure the classes get applied to the image elements
* `repos/keg-components/src/components/loading/loading.js`
  * Lots of clean up to fix how loading and spinners works
    * Spinners are now centered, and look much better

**Hooks** - Added a number of hooks for the className updates
* `repos/keg-components/src/hooks/useClassList.js`
  * Joins an array of class names together and memoizes them
* `repos/keg-components/src/hooks/useClassName.js`
  * Returns a function that should be set as the element ref
  * When the ref gets called, it gets the element and applies the `classNames` to the dom element
* `repos/keg-components/src/hooks/useThemePath.js`
  * Cleaned this up a lot
  * Now easier to understand, and seems to be faster
  * It's also doing a lot less
* `repos/keg-components/src/hooks/useThemeTypeAsClass.js`
  * Helper hook to use the `themePath`, and `type` props of a component to build the className
  * Calls `useClassList` internally

**Theme**
  * `repos/keg-components/src/theme/themeDefaults.js`
    * Updated theme defaults to allow being overwritten
    * Updated all references to `defaults.json` to use `getThemeDefaults` instead 
    * Added a `setThemeDefaults` that accepts a `defaults` object
    * **IMPORTANT** - This should be called prior to theme initialization
      * This way the defaults will be updated before the theme is built
  * `repos/keg-components/src/utils/styles/buildColorStyles.js`
    * Added `buildSurfaceStyles` helper to build styles based on defined `surfaces`
  * `repos/keg-components/src/theme/components/loading.js`
    * Updated how the styles are created
    * Uses the `buildSurfaceStyles`
  * 
**Keg-Core**
* `app.json`
  * Added mapping of `react-native-svg` to `react-native-svg-web` on web platform


## Testing
* Run `keg dpg run package=docker.pkg.github.com/simpleviewinc/keg-packages/keg-components:zr-354-classname-updates`
  * All components should have a `keg-*` class name applied to them
  * The `*` is based on the component name and location
  * So `keg-card-container` would be the container component with-in the card

**Steps to Test**
  * Open up the inspector in your browser
  * Then, starting with the first component; right click it an select `inspect`
  * Find the component within the `inspectors` dom tree view
  * Ensure that it has a class name applied to it
    * It should look like  [this](http://snpy.in/Y1WwtP)
  * Do this for all components and ensure they have a `keg-*` class applied to them
  * **IMPORTANT** - There are some cases `react-native-web` will generate child dom elements
    * These will not have the class applied to them
    * Technically I could have added them, doing something like what was done on the Image component
    * But it's possible for `react-native-web` to change what elements are created in future versions, which would break how this works.
    * I did do it for the image, because I felt it was important to have a class directly applied to images
      * Images tend to have more styles applied to them, which means they need a way to be accessed
      * You can see an example of the `react-native-web` generated elements [here](http://snpy.in/1IGHBI)
        * In that image, the element with the `keg-image-sibling` class is created by `react-native-web`; not `keg-components`
